File: code/swz/swz_figure1.jl
using MAT
using Plots
using StatsPlots
using LaTeXStrings

# Path for figures
figure_path = "/Users/felipe/Library/CloudStorage/OneDrive-UniversidadTorcuatoDiTella/Maeco/1. Primer Año/Tercer Trimestre/macro2 - Macroeconomía II/take-home/code/swz/figures"

# Create figures directory if it doesn't exist
mkpath(figure_path)

# Define colors for the curves
main_color = colorant"#084B83"
alt_color = colorant"#FF7F0E"

# Add plot defaults for fonts
default(
    tickfont=font(12, "Computer Modern"),
    guidefont=font(14, "Computer Modern"),
    legendfont=font(12, "Computer Modern")
)

# Read the MATLAB data
data = matread("/Users/felipe/Library/CloudStorage/OneDrive-UniversidadTorcuatoDiTella/Maeco/1. Primer Año/Tercer Trimestre/macro2 - Macroeconomía II/take-home/code/swz/model/irf_data.mat")
t = vec(data["t"])
T = data["T"]

# Extract IRF data
Yirf0, Yirf1 = data["Yirf0"], data["Yirf1"]
Xirf0, Xirf1 = data["Xirf0"], data["Xirf1"]
Piirf0, Piirf1 = data["Piirf0"], data["Piirf1"]
Rsirf0, Rsirf1 = data["Rsirf0"], data["Rsirf1"]

# Function to create a subplot with common styling
function make_subplot(data1, data2, ylabel, legend_position=:bottomright)
    p = plot(t, data1, 
        label=L"\textrm{Modelo\ NK\ 4\ ec.}", 
        color=main_color, 
        linestyle=:solid, 
        linewidth=2,
        grid=true,
        xlabel=L"\textrm{Horizonte}",
        ylabel=ylabel,
        xlims=(1, 21),
        xticks=(1:5:21, ["0", "5", "10", "15", "20"])
    )
    plot!(p, t, data2, 
        label=L"\textrm{Modelo\ NK\ 3\ ec.}", 
        color=alt_color, 
        linestyle=:dash, 
        linewidth=2
    )
    if legend_position != :none
        plot!(p, legend_position=legend_position)
    end
    return p
end

# Productivity / Natural Rate Shock
plt1 = plot(layout=(2,2), size=(800,600))
plot!(plt1[1], t, 100 .* Yirf0[:,1], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt1[1], t, 100 .* Yirf1[:,1], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt1[1], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Producto}", grid=true, legend=:none)

plot!(plt1[2], t, 100 .* Xirf0[:,1], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt1[2], t, 100 .* Xirf1[:,1], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt1[2], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Brecha\ de\ Producto}", grid=true, legend=:none)

plot!(plt1[3], t, 400 .* Piirf0[:,1], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt1[3], t, 400 .* Piirf1[:,1], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt1[3], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Inflación}", grid=true, legend=:none)

plot!(plt1[4], t, 400 .* Rsirf0[:,1], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt1[4], t, 400 .* Rsirf1[:,1], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt1[4], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Tasa\ de\ Interés}", grid=true, legend=:bottomright)

for i in 1:4
    plot!(plt1[i], xlims=(1,21), xticks=(1:5:21, ["0", "5", "10", "15", "20"]))
end

savefig(plt1, joinpath(figure_path, "four_nat.pdf"))

# Monetary Shock
plt2 = plot(layout=(2,2), size=(800,600))
plot!(plt2[1], t, 100 .* Yirf0[:,2], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt2[1], t, 100 .* Yirf1[:,2], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt2[1], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Producto}", grid=true, legend=:none)

plot!(plt2[2], t, 100 .* Xirf0[:,2], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt2[2], t, 100 .* Xirf1[:,2], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt2[2], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Brecha\ de\ Producto}", grid=true, legend=:none)

plot!(plt2[3], t, 400 .* Piirf0[:,2], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt2[3], t, 400 .* Piirf1[:,2], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt2[3], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Inflación}", grid=true, legend=:none)

plot!(plt2[4], t, 400 .* Rsirf0[:,2], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt2[4], t, 400 .* Rsirf1[:,2], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt2[4], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Tasa\ de\ Interés}", grid=true, legend=:bottomright)

for i in 1:4
    plot!(plt2[i], xlims=(1,21), xticks=(1:5:21, ["0", "5", "10", "15", "20"]))
end

savefig(plt2, joinpath(figure_path, "four_pol.pdf"))

# Leverage Shock
plt3 = plot(layout=(2,2), size=(800,600))
plot!(plt3[1], t, 100 .* Yirf0[:,4], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt3[1], t, 100 .* Yirf1[:,4], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt3[1], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Producto}", grid=true, legend=:none)

plot!(plt3[2], t, 100 .* Xirf0[:,4], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt3[2], t, 100 .* Xirf1[:,4], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt3[2], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Brecha\ de\ Producto}", grid=true, legend=:none)

plot!(plt3[3], t, 400 .* Piirf0[:,4], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt3[3], t, 400 .* Piirf1[:,4], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt3[3], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Inflación}", grid=true, legend=:none)

plot!(plt3[4], t, 400 .* Rsirf0[:,4], label=L"\textrm{Modelo\ NK\ 4\ ec.}", color=main_color, linestyle=:solid, linewidth=2)
plot!(plt3[4], t, 400 .* Rsirf1[:,4], label=L"\textrm{Modelo\ NK\ 3\ ec.}", color=alt_color, linestyle=:dash, linewidth=2)
plot!(plt3[4], xlabel=L"\textrm{Horizonte}", ylabel=L"\textrm{Tasa\ de\ Interés}", grid=true, legend=:topright)

for i in 1:4
    plot!(plt3[i], xlims=(1,21), xticks=(1:5:21, ["0", "5", "10", "15", "20"]))
end
savefig(plt3, joinpath(figure_path, "four_lev_qe.pdf"))

File: code/swz/swz_figure1.m
% Author: Felipe Tappata
% Code adapted from Sims, Wu and Zhang (2019) "The Four Equation New Keynesian Model"
% Original code downloaded from: https://sites.google.com/view/jingcynthiawu/

clear
close all
cd('/Users/felipe/Library/CloudStorage/OneDrive-UniversidadTorcuatoDiTella/Maeco/1. Primer Año/Tercer Trimestre/macro2 - Macroeconomía II/take-home/code/swz/model')

beta = 0.995; % discount rate patient
Pis = 1; % trend inflation
Rss = Pis/beta; % SS short rate
Rres = Rss; % SS rate on reserves
sp = 2/400; % targeted spread
Rbs = sp + Rss; % steady state long rate
betab = beta*Rss/Rbs; % discount factor impatient
epsi = 11; % elasticity of sub
pms = (epsi-1)/epsi; % steady state markup
ws = pms; % steady state wage
Ls = 1; % targeted SS labor
phi = 0.75; % calvo parameter
prs = ( ( 1-phi*Pis^(epsi-1) )/(1-phi))^(1/(1-epsi)); % SS reset price
vs = (1-phi)*prs^(-epsi)/(1-phi*Pis^(epsi)); % SS dispersion
Ys = Ls/vs; % SS output
z = 0.33; % Targeted share of impatient types
Cbs = z*Ys; % SS consumption impatient
Cs = (1-z)*Ys; % SS consumption patient
chi = 1; % Inverse Frisch
sigma = 1; % EIS
psi = Cs^(-sigma)*ws/Ls^(chi); % Labor disutility consistent with 
r = 0.1; % SS reserve/output ratio
res = r*Ys; % SS reserves
kappa = 1 - 40^(-1); % Perpetuity decay
Qs = (Rbs - kappa)^(-1); % Steady state bond price
bcbs = res/Qs; % steady state CB bond holdings
bs = Cbs/Qs;
Xb = (Rbs*Qs*bs);
bfis = bs - bcbs; % Steady state FI bond holdings
thetas = 5; % modified leverage ratio
Xfi = Qs*bfis/thetas; % Transfer to FI
ss = Qs*bfis*(1-kappa*Pis^(-1)) + res - Xfi; % steady state deposits/saving
qes = Qs*bcbs;

% checks
Ds = Ys - ws*Ls; % production dividends
Dfi = Rbs*Qs*bfis + Rres*res - Rss*ss; % FI dividends
Ts = -Rbs*Qs*bcbs + Rres*res; % lump sum transfer (should be negative because it is a transfer)

check = Cs + ss - ws*Ls - Rss*ss - Ds - Dfi + Ts + kappa*Qs*bfis + Xfi + Xb; % Should be zero -- patient household budget constraint

% Taylor rule
rhor = 0.8;
phipi = 1.5;
phix = 0.0;

% Exogenous processes
rhoA = 0.8;
rhoq = 0.8;
rhot = 0.8;

srf = 1/400; % desired shock to natural rate. Pick sA to generate this -- 100 basis points annualized
sA = abs(.01*(chi*(1-z) + sigma)/((1+chi)*(1-z)));
sq = 0.01;
st = 0.01;
sr = 0.01;

% Save parameters
gamma = (1-phi)*(1-phi*beta)/phi;
zeta = (chi*(1-z)+sigma)/(1-z);
b1 = bfis/bs;
b2 = bcbs/bs;

save param_4nk beta betab z psi chi sigma kappa Xfi Xb thetas epsi phi bcbs phipi phix Pis rhoq rhoA rhot rhor sq sA st sr Rres Rbs Qs bfis res ss Rbs Rss sp bcbs bs Cbs qes gamma zeta b1 b2

% Four equation model
dynare model_4nk noclearall nolog

% Calculate conversions and IRFs
xr = oo_.irfs.Y_eA(1)/oo_.irfs.Y_er(1);
xq = oo_.irfs.Y_eA(1)/oo_.irfs.Y_eq(1);
xt = oo_.irfs.Y_eA(1)/oo_.irfs.Y_et(1);
check1 = oo_.irfs.Yf_eA(1);
Yirf0 = [oo_.irfs.Y_eA' xr*oo_.irfs.Y_er' xq*oo_.irfs.Y_eq' xt*oo_.irfs.Y_et'];
Piirf0 = [oo_.irfs.Pi_eA' xr*oo_.irfs.Pi_er' xq*oo_.irfs.Pi_eq' xt*oo_.irfs.Pi_et'];
Xirf0 = [oo_.irfs.X_eA' xr*oo_.irfs.X_er' xq*oo_.irfs.X_eq' xt*oo_.irfs.X_et'];
Rsirf0 = [oo_.irfs.Rs_eA' xr*oo_.irfs.Rs_er' xq*oo_.irfs.Rs_eq' xt*oo_.irfs.Rs_et'];
T_et = xt*oo_.irfs.theta_et;
QE_eq = xq*oo_.irfs.qe_eq;

% Three equation model
sA = abs(.01*(chi*(1-0) + sigma)/((1+chi)*(1-0)));
zeta = (chi*(1-0)+sigma)/(1-0);
save param_3nk beta betab z psi chi sigma kappa Xfi Xb thetas epsi phi bcbs phipi phix Pis rhoq rhoA rhot rhor sq sA st sr Rres Rbs Qs bfis res ss Rbs Rss sp bcbs bs Cbs qes gamma zeta b1 b2
dynare model_3nk noclearall nolog
Yirf1 = [oo_.irfs.Y_eA' xr*oo_.irfs.Y_er' xq*oo_.irfs.Y_eq' xt*oo_.irfs.Y_et'];
Piirf1 = [oo_.irfs.Pi_eA' xr*oo_.irfs.Pi_er' xq*oo_.irfs.Pi_eq' xt*oo_.irfs.Pi_et'];
Xirf1 = [oo_.irfs.X_eA' xr*oo_.irfs.X_er' xq*oo_.irfs.X_eq' xt*oo_.irfs.X_et'];
Rsirf1 = [oo_.irfs.Rs_eA' xr*oo_.irfs.Rs_er' xq*oo_.irfs.Rs_eq' xt*oo_.irfs.Rs_et'];
check2 = oo_.irfs.Yf_eA(1);

T = size(Yirf0,1);
t = 1:T;

% Save IRF data
save('irf_data.mat', 'Yirf0', 'Yirf1', 'Xirf0', 'Xirf1', 'Piirf0', 'Piirf1', 'Rsirf0', 'Rsirf1', 't', 'T');
